Trimmed block-start margins for flex items in horizontal writing mode should be reflected in computed style

sammygill opened the pull request:
#### 81223299d3fb4a43229294386028092f559b5944
<pre>
Trimmed block-start margins for flex items in horizontal writing mode should be reflected in computed style
<a href="https://bugs.webkit.org/show_bug.cgi?id=253712">https://bugs.webkit.org/show_bug.cgi?id=253712</a>
rdar://106558991

Reviewed by Alan Baradlay.

Whenever we trim a block-start margin for a flex item, we should include
a bit of information for this trimmed margin as a piece of rare data
for the object. The trimming is done inside trimMainAxisMarginStart and
trimCrossAxisMarginStart, so we can wrap abstract the trimming into a
new helper method, RenderBlock::setTrimmedMarginForChild. This new
helper will trim the correct margin and also set the corresponding bit
of rare data

Then, inside of ComputedStyleExtractor we can check
if this bit has been set when trying to determine the block-start margin
when the flexbox is in horizontal writing-mode (i.e. the top margin).

For flex layout, we can clear these new bits whenever we go through
another layout and recreate the flex items.

We also need to update the conditions for determining whether the
margin-top property is layout dependent by taking into consideration
margin-trim. For flex items, this can be done by checking the set
margin-trim values on the flexbox.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-multi-line-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-multi-line-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start-dynamic-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start-dynamic.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-multi-line-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-multi-line-block-start.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::isFlexItem):
(WebCore::rendererContainingBlockHasMarginTrim):
(WebCore::isLayoutDependent):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::setTrimmedMarginForChild):
* Source/WebCore/rendering/RenderBlock.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::physicalToFlowRelativeDirectionMapping const):
(WebCore::flowRelativeDirectionToMarginTrimType):
(WebCore::RenderBox::markMarginAsTrimmed):
(WebCore::RenderBox::clearTrimmedMarginsMarkings):
(WebCore::RenderBox::hasTrimmedMargin const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::trimMainAxisMarginStart):
(WebCore::RenderFlexibleBox::trimCrossAxisMarginStart):
(WebCore::RenderFlexibleBox::constructFlexItem):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: <a href="https://commits.webkit.org/262081@main">https://commits.webkit.org/262081@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aea90da13723819d5f8e62382090c7858be3b04c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/309 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/325 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/339 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/316 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/294 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/311 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/404 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/511 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/552 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/316 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/351 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/341 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/388 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/299 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/377 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/275 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/316 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/298 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/268 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/308 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/120 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/301 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-multi-column-block-start-expected.txt:
```diff
@@ -0,0 +1,6 @@
+
```
multi-column name is "reserved" for fragmentation, please do not use it in the context of flex layout (it's highly misleading as people would immediately assume column-count/width values here and not just a multi-line flex container with column/flow direction)

sammygill replied:
Changed file names

alanbaradlay commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2285,6 +2286,27 @@ static inline bool isNonReplacedInline(RenderObject& renderer)
     return renderer.isInline() && !renderer.isReplacedOrInlineBlock();
 }
 
+static bool isFlexItem(RenderObject* renderer)
```
const RenderObject

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2285,6 +2286,27 @@ static inline bool isNonReplacedInline(RenderObject& renderer)
+{
+    ASSERT(renderer.isFlexItem());
+    auto* containingBlock = renderer.containingBlock();
+    if (is<HTMLHtmlElement>(renderer.element()) || containingBlock->isRenderView())
```
Any reason of mixing DOM and render tree check? I think it should be able to keep this check in either domain.

alanbaradlay replied:
what guarantees a non-null containing block?

sammygill replied:
Fixed both issues (kept the render tree check)

alanbaradlay commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2285,6 +2286,27 @@ static inline bool isNonReplacedInline(RenderObject& renderer)
+
+static bool rendererContainingBlockHasMarginTrim(const RenderBox& renderer, std::optional<MarginTrimType> marginTrimType)
+{
+    ASSERT(renderer.isFlexItem());
```
If we expect the incoming renderer to be a flex item, I think this function should be name accordingly as opposed to using some generic terms - as what we seem to be checking here is whether the flex container has margin trim.

alanbaradlay replied:
(unless this is going to evolve into a more generic function, in which case I'd not assert against the type but rather indicate that we don't support other cases yet (like in the other trim function below).

sammygill replied:
Yes, the idea is this to be generic and will eventually check all 3 layout types for margin-trim. Modified the logic to assert not implemented for the layout types this isn't implemented for yet (block layout and grid)

alanbaradlay commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2721,6 +2744,27 @@ bool ComputedStyleExtractor::hasProperty(CSSPropertyID propertyID)
     return propertyValue(propertyID);
 }
 
+static bool rendererHasTrimmedMargin(const RenderElement* renderer, CSSPropertyID propertyId)
```
marginTrim vs. trimmedMargin is going to be a bit confusing. I assume marginTrim means computed value while trimmedMargin means layout output, is that correct?

sammygill replied:
`trimmedMargin` is supposed to indicate that the box has a trimmed margin (not sure if that matches with the definition of layout output). Perhaps a better name would be `marginTrimMarkings` since the vocabulary that I have been using is that we "mark," bits

alanbaradlay replied:
I am not sure if markings makes it any clearer since the term "markings" does not really indicate a transition from "computed" to "result of layout" but again I am not good at namings. 🤷 
I wouldn't be too worried about this though at this point.

alanbaradlay commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2721,6 +2744,27 @@ bool ComputedStyleExtractor::hasProperty(CSSPropertyID propertyID)
+        ASSERT_NOT_REACHED();
+        return false;
+    }
+    return false;
```
I believe this line is redundant.

sammygill replied:
Removed!

alanbaradlay commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -3251,7 +3295,7 @@ RefPtr<CSSValue> ComputedStyleExtractor::valueForPropertyInStyle(const RenderSty
         return CSSPrimitiveValue::createCustomIdent(style.specifiedLocale());
     case CSSPropertyMarginTop:
-        return zoomAdjustedPaddingOrMarginPixelValue<&RenderStyle::marginTop, &RenderBoxModelObject::marginTop>(style, renderer);
+        return rendererHasTrimmedMargin(renderer, propertyID) ? zoomAdjustedPixelValue(downcast<RenderBox>(*renderer).marginTop(), style) : zoomAdjustedPaddingOrMarginPixelValue<&RenderStyle::marginTop, &RenderBoxModelObject::marginTop>(style, renderer);
```
This line assumes the only RenderBox type of renderers can have trimmed margins. While this is currently true, a change in rendererHasTrimmedMargin can result in incorrect downcast (not a big deal, but  in cases like this I'd refactor the code so that we don't have to make such assumptions)

sammygill replied:
Made a couple of changes here that hopefully line up with what you were thinking about. 

1.) Tried to dynamicDowncast to a RenderBox and continue with the logic only if it works
2.) Change `rendererHasTrimmedMargin` to take a RenderBox as an argument. 
So this should still work unless the function's argument type significantly changes I believe

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1394,6 +1394,75 @@ void RenderBox::clearOverridingLogicalWidthLength()
+{
+    if (!hasRareData())
+        return;
+    markMarginAsTrimmed(static_cast<MarginTrimType>(0));
```
how does this clear the trimmed margin off of the rare data (markMarginAsTrimmed uses bitwise or).

sammygill replied:
Changed it to just set the marking to 0 directly here

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1394,6 +1394,75 @@ void RenderBox::clearOverridingLogicalWidthLength()
+
+bool RenderBox::hasTrimmedMargin(PhysicalDirection physicalDirection) const
+{
+    ASSERT(isFlexItem());
```
(same as above)

sammygill replied:
Added logic to assert when called with unsupported layout type

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1394,6 +1394,75 @@ void RenderBox::clearOverridingLogicalWidthLength()
 
+std::optional<FlowRelativeDirection> RenderBox::physicalToFlowRelativeDirectionMapping(PhysicalDirection direction, LayoutCalculationPhase layoutCalculationPhase) const
+{
+    // FIXME: Return type should be just FlowRelativeDirection once all code paths have been implemented
```
I don't think the lack of supported codepaths in this function should drive the return type which means you need to adjust all the callsites as you progress your implementation. We should always be able to resolve physical to relative. If you end up on an unsupported codepath, please assert and return the default value.

sammygill replied:
Yeah I wasn't entirely sure what a good default value would be which is why I added the optional. Changed that and added some asserts. I like the skeleton code structure here to highlight the exact logic that still needs to be added for it to be considered complete, but we can condense it into one assert if you would like

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1670,6 +1670,9 @@ FlexItem RenderFlexibleBox::constructFlexItem(RenderBox& child, bool relayoutChi
         downcast<RenderFlexibleBox>(child).resetHasDefiniteHeight();
+
+    if (childHadLayout && !style().marginTrim().isEmpty())
+        child.clearTrimmedMarginsMarkings();
```
How do you handle the case when the flex container transitions from margin-trim to non-margin-trim?

sammygill replied:
Changed to check if the child has any markings instead

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -47,6 +47,12 @@ enum ShouldComputePreferred { ComputeActual, ComputePreferred };
 enum class StretchingMode { Any, Explicit };
 
+enum class PhysicalDirection { Top, Bottom, Left, Right };
+enum class FlowRelativeDirection { BlockStart, BlockEnd, InlineStart, InlineEnd };
```
I'd look into LogicalBoxSide/BoxSide enums to see if you could reuse them.

sammygill replied:
Changed to create a type alias with `using`

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1394,6 +1394,75 @@ void RenderBox::clearOverridingLogicalWidthLength()
+bool RenderBox::hasTrimmedMargin(PhysicalDirection physicalDirection) const
+{
+    ASSERT(isFlexItem());
+    auto flowRelativeDirection = physicalToFlowRelativeDirectionMapping(physicalDirection, LayoutCalculationPhase::Positioning);
```
How do you enforce this function being called only during Positioning phase? Currently the only callsite I see is ComputedStyleExtractor::valueForPropertyInStyle() which I can't tell if belongs to Positioning or not.

sammygill replied:
This is interesting because I'm not actually sure if it's possible to check which phase we are in within render tree is there (i.e. we don't keep track of such a construct during layout)?

For now I added an assert to make sure this is only called on a box that does not need to go through layout since it doesn't really make sure to check for trimmed margins otherwise. I *think* that's a reasonable assertion in this case but let me know if you think otherwise...

alanbaradlay commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -2721,6 +2750,22 @@ bool ComputedStyleExtractor::hasProperty(CSSPropertyID propertyID)
     return propertyValue(propertyID);
 }
 
+static bool rendererHasTrimmedMargin(const RenderBox& renderer, CSSPropertyID propertyId)
```
nit: hasTrimmedMargin and containingBlockHasMarginTrim should just work.

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1394,6 +1394,87 @@ void RenderBox::clearOverridingLogicalWidthLength()
+    auto isHorizontalWritingMode = style().isHorizontalWritingMode();
+    auto isLeftToRightDirection = style().isLeftToRightDirection();
+
+    if (layoutCalculationPhase == LayoutCalculationPhase::Positioning) {
```
I don't think you should introduce new terms unless there's a clear path forward to enforce them which I don't believe there's given how our layout codebase is structured at the moment.

sammygill replied:
Removed

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1670,6 +1670,9 @@ FlexItem RenderFlexibleBox::constructFlexItem(RenderBox& child, bool relayoutChi
     if (is<RenderFlexibleBox>(child))
         downcast<RenderFlexibleBox>(child).resetHasDefiniteHeight();
+
+    if (childHadLayout && child.hasTrimmedMargin(std::optional<MarginTrimType> { }))
```
child.hasTrimmedMargin({ }) ?

sammygill replied:
Yeah I had that, but unfortunately it was calling the wrong `hasTrimmedMargin `

alanbaradlay commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -939,6 +956,8 @@ inline void RenderBox::setInlineBoxWrapper(LegacyInlineElementBox* boxWrapper)
 
 LayoutUnit synthesizedBaseline(const RenderBox&, const RenderStyle& parentStyle, LineDirectionMode, BaselineSynthesisEdge);
 
+MarginTrimType flowRelativeDirectionToMarginTrimType(FlowRelativeDirection);
```
No need to expose this to .h at this point.

sammygill replied:
Removed

alanbaradlay reviewed (approved):
Please add dynamic test cases covering bugs found at code review.

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1394,6 +1394,85 @@ void RenderBox::clearOverridingLogicalWidthLength()
+{
+    ASSERT(hasRareData());
+    ensureRareData().setTrimmedMargins(0);
+    setNeedsLayout(MarkOnlyThis);
```
Could you explain the logic behind this call, please? To my understanding clearTrimmedMarginsMarkings() is about invalidating a cached value of whether the associated box has trimmed margins, right? This value is valid from the point of computing the trimmed value (if applicable) until we believe the renderer needs new geometry (for whatever reason). It makes me believe that at the point of the invalidation the associated box should already be marked damaged and if not, the callsite should be doing that instead (it has more context to decide the extent of the damage as calling setNeedsLayout only on *this does not change the flow of the layout e.g. the renderer may very well stay damaged)

sammygill replied:
My idea is that these 2 things (unmarking the trimming bits on a child and laying it out again) would always go together and am not sure why clearing the bits would not be followed with a layout. However, I do see you point about the caller having more context about this and my idea was that it would only be used in those scenarios where the caller was attempting to perform a layout (probably because it is involved with some sort of formatting context), of some children and clearing the bits is an indication that a box needs to go through layout again.

I could also see how someone could accidentally call this method in a scenario where they probably don't mean to and find themselves with a box that was marked for layout. 

I went ahead and removed the call from the function and moved it to the call site of `clearTrimmedMarginsMarkings ` instead.

alanbaradlay replied:
We normally mark renderers dirty when
1, some kind of style or content mutation happens (before layout)
2, need to expand the damage on descendant renders (or siblings/nephews etc in rare cases) during layout because constraints etc change and we believe it may have impact on other renderers.
(The second needs carful handling because layout has a well defined flow and as I said earlier marking renderers dirty has no impact on the direction of the flow)
Trimmed margin on a flex item itself should not trigger layout. If margin changes on the child, yes we need to invalidated this value (child is already marked dirty), if flex container's margin-trim value changes, yes we need to invalidate and mark flex items dirty) etc. 
I think the dependency is the other way around, i.e. when the child needs layout, the cached value needs to be invalidated as well.
(does this need to be addressed in this PR?)

sammygill replied:
That makes sense I appreciate the thorough explanation on this. I went ahead and removed the call to `setChildNeedsLayout`.

I think that the patch is ok the way it is now for the scenario where we need to unmark the children when they are dirty because:
1.) When the flex box constructs the flex items again in `RenderFlexibleBox::constructFlexItem` we will unmark the child if it has the bits set
2.) As the flexbox goes through and recreates the lines `flexAlgorithm.computeNextFlexLine` we will set the bits for each main axis edge appropriately
3.) As it sizes each flex line in the cross axis, we will trim each edge when needed 

That happens regardless if a flex item renderer actually needs to go through layout again or not so it should be ok.

As for marking the flex items when the flex box's margin-trim value changes, I was planning on addressing that at the same time I make a margin-trim addition to `rareDataChangeRequiresLayout`.

alanbaradlay replied:
Alternatively you can introduce an invalid state and compute the cached value lazily on the first call.

alanbaradlay reviewed (changes requested):
(setNeedsLayout needs to be addressed before landing)

alanbaradlay commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1394,6 +1394,84 @@ void RenderBox::clearOverridingLogicalWidthLength()
 
+FlowRelativeDirection RenderBox::physicalToFlowRelativeDirectionMapping(PhysicalDirection direction) const
+{
+    auto isHorizontalWritingMode = style().isHorizontalWritingMode();
+    auto isLeftToRightDirection = style().isLeftToRightDirection();
+
+    if (isHorizontalWritingMode == containingBlock()->style().isHorizontalWritingMode()) {
+        switch (direction) {
+        case PhysicalDirection::Top:
+            return isHorizontalWritingMode ? FlowRelativeDirection::BlockStart : (isLeftToRightDirection ? FlowRelativeDirection::InlineStart : FlowRelativeDirection::InlineEnd);
```
At some point this function needs to be revisited and figure out what this mapping really means since if you try to map directions on *this, it's actually never based on *this style, but the context *this is in (parent, containing block etc depending on the layout system) for example, an inline-block with trimmed-margin can have "direction: rtl" but the direction applies to its _content_ and not to the inline-block as an inline level box (i.e. it does not make the inline level box (initiated by the inline-block) suddenly "right direction" on the line).

sammygill committed:
Trimmed block-start margins for flex items in horizontal writing mode should be reflected in computed style
https://bugs.webkit.org/show_bug.cgi?id=253712
rdar://106558991

Reviewed by Alan Baradlay.

Whenever we trim a block-start margin for a flex item, we should include
a bit of information for this trimmed margin as a piece of rare data
for the object. The trimming is done inside trimMainAxisMarginStart and
trimCrossAxisMarginStart, so we can wrap abstract the trimming into a
new helper method, RenderBlock::setTrimmedMarginForChild. This new
helper will trim the correct margin and also set the corresponding bit
of rare data

Then, inside of ComputedStyleExtractor we can check
if this bit has been set when trying to determine the block-start margin
when the flexbox is in horizontal writing-mode (i.e. the top margin).

For flex layout, we can clear these new bits whenever we go through
another layout and recreate the flex items.

We also need to update the conditions for determining whether the
margin-top property is layout dependent by taking into consideration
margin-trim. For flex items, this can be done by checking the set
margin-trim values on the flexbox.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-multi-line-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-column-multi-line-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start-dynamic-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start-dynamic.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-block-start.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-multi-line-block-start-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/computed-margin-values/flexbox-row-multi-line-block-start.html: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::isFlexItem):
(WebCore::rendererContainingBlockHasMarginTrim):
(WebCore::isLayoutDependent):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::setTrimmedMarginForChild):
* Source/WebCore/rendering/RenderBlock.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::physicalToFlowRelativeDirectionMapping const):
(WebCore::flowRelativeDirectionToMarginTrimType):
(WebCore::RenderBox::markMarginAsTrimmed):
(WebCore::RenderBox::clearTrimmedMarginsMarkings):
(WebCore::RenderBox::hasTrimmedMargin const):
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::trimMainAxisMarginStart):
(WebCore::RenderFlexibleBox::trimCrossAxisMarginStart):
(WebCore::RenderFlexibleBox::constructFlexItem):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: https://commits.webkit.org/262081@main
